auth/authzn.proto

```diff
@@ -0,0 +1,179 @@
+    // https://grpc.io/docs/guides/error/ for more details.
+    rpc Set(SetRequest) returns (SetResponse) {}
+}
+
+message SetRequest {
+    // A list of policies that are to be applied on the network element.
+    // Multiple policies per service MAY be specified to allow for multiple
+    // users, or specific user restrictions to be described.
+    repeated Policy policy = 1;
+}
+
+// Policy specifies an authentication/authorization policy for an individual
+// service that is instantiated on the device. Some services may only have an
+// authentication policy - where they do not support more granular authorization,
+// whereas others may have both specified.
+message Policy {
+    oneof protocol {
+        // console provides the configuration for authentication of the local
+        // system login.
+        Console console = 1;
+        // ssh provides the configuration for authentication of logins to the
+        // system via SSH.
+        SSH ssh = 2;
+        // gnmi provides a local authentication mechanism for all RPCs within
+        // the gNMI service, along with an authorization policy which allows
+        // users to be restricted to:
+        //  - specific RPCs (e.g., Get, Set, Subscribe, Capabilities)
+        //  - specific restrictions based on path within each RPC.
+        gNMI gnmi = 3;
+        // gnoi provides a local authentication mechanism for each gNOI RPC
+        // within the suite of gNOI microservices whose definitions are
+        // defined in github.com/openconfig/gnoi. Within specific microservices
+        // RPC-based restrictions may be realised.
+        gNOI gnoi = 4;
+        // gribi provides a local authentication mechanism for the gRIBI service.
+        gRIBI gribi = 5;
+    }
+}
+
+message Console {
+    // Console provides a username/password based authentication mechanism
+    // for logging in to the system. 
+    //
+    // The username of the user permitted to log in to the system via the 
+    // console. If the username is specified as an
+    // empty string, the supplied authentication details should apply to the
+    // root/superuser/admin user of the system.
+    string username = 1;
+    // TODO(robjs): How should this password be supplied to the device?
+    // Presumably we do not want it in cleartext. We could re-use the same
+    // definition that is used within the OpenConfig models:
+    // http://ops.openconfig.net/branches/models/master/docs/openconfig-system.html#type-crypt-password-type
+    string password = 2;
+}
+
+message SSH {
+    // SSH provides a key-based login for users onto the system. Only users
+    // that are specified within the SSH policy are permitted to login to the
+    // system.
+    //
+    // The username of a user for whom SSH login is permitted. If the value
+    // of the username field is an empty string, this corresponds to the
+    // root/superuser/admin user of the system.
+    string username = 1;
```

Comment:
If the user specified in the 'username' field does not exist on the box at the moment should it be created? If so, what level of access it should be given?

Reply:
What should be done if an entry with username=="" is not present on the list? (I assume the root/admin account should not be removed)

Reply:
For SSH access, what is the granularity of user access that you are considering? I was assuming that we are thinking that this is either allowed or not allowed - nothing more granular than this?

w.r.t to no `username=""` - I would assume that this translates to `PermitRootLogin no` in OpenSSH's `sshd_config` - rather than removing the user?

Reply:
That still leaves question what to do if a username is not known to the device. Should the user be created? Should the whole request be rejected?